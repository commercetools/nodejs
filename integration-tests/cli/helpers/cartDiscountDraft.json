{
  "name": {
    "en": "integration-test-cart-discount"
  },
  "value": {
    "type": "relative",
    "permyriad": 1000
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "target": {
    "type": "lineItems",
    "predicate": "1 = 1"
  },
  "sortOrder": "0.1",
  "requiresDiscountCode": true
}
