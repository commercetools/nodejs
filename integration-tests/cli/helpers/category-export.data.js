export default [
  {
    key: 'fooKey',
    name: {
      en: 'categoryA',
    },
    slug: {
      en: 'category-a',
    },
    description: {
      en: 'descA',
    },
    externalId: 'a',
    assets: [],
  },
  {
    key: 'barKey',
    name: {
      en: 'categoryB',
    },
    slug: {
      en: 'category-b',
    },
    description: {
      en: 'descB',
    },
    externalId: 'b',
    assets: [],
  },
  {
    key: 'fooBarKey',
    name: {
      en: 'categoryC',
    },
    slug: {
      en: 'category-c',
    },
    description: {
      en: 'descC',
    },
    externalId: 'c',
    assets: [],
  },
]
