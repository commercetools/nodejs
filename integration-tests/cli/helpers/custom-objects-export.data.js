export default [
  {
    container: 'Ludus',
    key: 'copperKey',
    value: {
      paymentMethod: 'Cash',
      paymentID: '1',
    },
  },
  {
    container: 'Frobozz',
    key: 'jadeKey',
    value: {
      paymentMethod: 'cc',
      paymentID: '2',
    },
  },
  {
    container: 'Syrinx',
    key: 'crystalKey',
    value: {
      paymentMethod: 'new',
      paymentID: '3',
    },
  },
]
