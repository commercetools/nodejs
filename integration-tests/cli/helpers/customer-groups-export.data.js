export default [
  {
    groupName: 'Ludus',
    key: 'copperKey',
  },
  {
    groupName: 'Frobozz',
  },
  {
    groupName: 'Syrinx',
    key: 'crystalKey',
  },
]
