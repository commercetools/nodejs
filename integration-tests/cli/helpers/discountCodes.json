[{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "IT8MKKZX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlN85n0"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDHzsVX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlb6Z1u"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITfWgUVd"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITpQgE1P"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITL2pvbZ"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlrnkj5"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDJO3LT"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITuIcDI2"
}]
