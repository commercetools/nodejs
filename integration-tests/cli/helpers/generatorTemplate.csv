name.en,name.de,description.en,description.de,cartPredicate,isActive,maxApplications,maxApplicationsPerCustomer
Sammy,Valerian,greatest promo,super angebot,some cart predicate,true,9,4
