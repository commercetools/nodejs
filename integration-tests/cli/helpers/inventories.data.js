export const inventories = [{
  sku: '12345',
  quantityOnStock: 20,
  custom: {
    type: {
      key: 'custom-type',
    },
    fields: {
      description: 'integration tests!! arrgggh',
    },
  },
}]

export const customFields = [{
  key: 'custom-type',
  name: {
    en: 'customized fields',
  },
  description: {
    en: 'custom description',
  },
  resourceTypeIds: ['inventory-entry'],
  fieldDefinitions: [{
    name: 'description',
    type: {
      name: 'String',
    },
    required: false,
    label: {
      en: 'owner',
    },
    inputHint: 'SingleLine',
  }],
}]
