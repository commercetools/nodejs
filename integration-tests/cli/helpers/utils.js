import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import fetch from 'node-fetch'

export function clearData(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }
  return client.process(request, payload => {
    // Built-in states cannot be deleted
    const results =
      entityName === 'states'
        ? payload.body.results.filter(state => state.builtIn === false)
        : payload.body.results
    return Promise.all(
      results.map(result =>
        client.execute({
          uri: service
            .byId(result.id)
            .withVersion(result.version)
            .build(),
          method: 'DELETE',
        })
      )
    )
  })
}

export function createData(apiConfig, entityName, data, id) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })
  const requestOption = { projectKey: apiConfig.projectKey }
  const service = createRequestBuilder(requestOption)[entityName]
  return Promise.all(
    data.map(_data => {
      if (id) service.byId(id)
      const request = {
        uri: service.build(),
        method: 'POST',
        body: _data,
      }
      return client.execute(request)
    })
  )
}

export function getId(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }

  return client
    .execute(request)
    .then(result => Promise.resolve(result.body.results[0].id))
}
