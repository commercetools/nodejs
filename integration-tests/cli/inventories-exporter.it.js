import { getCredentials } from '@commercetools/get-credentials'
import { exec } from 'mz/child_process'
import fs from 'mz/fs'
import tmp from 'tmp'
import CONSTANTS from '@commercetools/inventories-exporter/lib/constants'
import { version } from '@commercetools/inventories-exporter/package.json'
import { inventories, customFields } from './helpers/inventories.data'
import { clearData, createData } from './helpers/utils'

let projectKey
if (process.env.CI === 'true')
  projectKey = 'inventories-export-integration-test'
else projectKey = process.env.npm_config_projectkey

describe('StockExporter CLI', () => {
  const binPath = './integration-tests/node_modules/.bin/inventoriesexporter'
  let apiConfig
  beforeAll(async () => {
    const credentials = await getCredentials(projectKey)
    apiConfig = {
      host: CONSTANTS.host.auth,
      apiUrl: CONSTANTS.host.api,
      projectKey,
      credentials: {
        clientId: credentials.clientId,
        clientSecret: credentials.clientSecret,
      },
    }
    await clearData(apiConfig, 'orders')
    await clearData(apiConfig, 'inventory')
    await clearData(apiConfig, 'types')
    await createData(apiConfig, 'types', customFields)
    await createData(apiConfig, 'inventory', inventories)
  }, 30000)

  afterAll(() =>
    clearData(apiConfig, 'inventory').then(() => clearData(apiConfig, 'types')))

  describe('CLI basic functionality', () => {
    it('should print usage information given the help flag', async () => {
      const [stdout, stderr] = await exec(`${binPath} --help`)
      expect(stdout).toMatchSnapshot()
      expect(stderr).toBeFalsy()
    })

    it('should print the module version given the version flag', async () => {
      const [stdout, stderr] = await exec(`${binPath} --version`)
      expect(stdout).toBe(`${version}\n`)
      expect(stderr).toBeFalsy()
    })

    it('should export inventories to file as json', async () => {
      const jsonFilePath = tmp.fileSync().name
      const [stdout, stderr] = await exec(
        `${binPath} -p ${projectKey} -o ${jsonFilePath}`
      )
      expect(stdout).toMatchSnapshot()
      expect(stderr).toBeFalsy()
      const data = await fs.readFile(jsonFilePath, { encoding: 'utf8' })
      const result = JSON.parse(data)
      expect(result.length).toBe(1)
      expect(result[0].sku).toBe(inventories[0].sku)
    })

    it('should accept query', async () => {
      const jsonFilePath = tmp.fileSync().name
      const queryFlag = '-q "sku=\\"invalid\\""'
      const [stdout, stderr] = await exec(
        `${binPath} -p ${projectKey} -o ${jsonFilePath} ${queryFlag}`
      )
      expect(stdout).toMatchSnapshot()
      expect(stderr).toBeFalsy()
      const data = await fs.readFile(jsonFilePath, { encoding: 'utf8' })
      expect(JSON.parse(data)).toEqual([])
    })

    it('should export inventories to file as csv', async () => {
      const csvFilePath = tmp.fileSync().name
      const [stdout, stderr] = await exec(
        `${binPath} -p ${projectKey} -o ${csvFilePath} -f csv`
      )
      expect(stdout && stderr).toBeFalsy()
      const data = await fs.readFile(csvFilePath, { encoding: 'utf8' })
      expect(data).toMatchSnapshot()
    })
  })
})
