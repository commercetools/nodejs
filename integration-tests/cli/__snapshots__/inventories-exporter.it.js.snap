// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockExporter CLI CLI basic functionality should accept query 1`] = `
"info Done exporting inventories.
"
`;

exports[`StockExporter CLI CLI basic functionality should export inventories to file as json 1`] = `
"info Done exporting inventories.
"
`;

exports[`StockExporter CLI CLI basic functionality should print usage information given the help flag 1`] = `
"

Usage: inventoriesexporter [options]
Exports inventories from the commercetools platform.

Options:
      --help         Show help                                         [boolean]
      --version      Show version number                               [boolean]
  -o, --outputFile   Path to output file.                    [default: \\"stdout\\"]
      --apiUrl       The host URL of the HTTP API service.
                     [default: \\"https://api.europe-west1.gcp.commercetools.com\\"]
      --authUrl      The host URL of the OAuth API service.
                    [default: \\"https://auth.europe-west1.gcp.commercetools.com\\"]
  -d, --delimiter    Used CSV delimiter.                          [default: \\",\\"]
      --accessToken  CTP client access token
  -p, --projectKey   API project key.                                 [required]
  -c, --channelKey   Channel key to use as filter for result to export.
                     Useful if you only have channel key but not id.
                     Can be used with the query flag

  -q, --query        Filter query for inventories:
                     dev.commercetools.com/http-api-projects-inventory.html#que
                     ry-inventory
                     can be used with channelKey flag

  -f, --format       Format for export[choices: \\"csv\\", \\"json\\"] [default: \\"json\\"]
  -t, --template     Path to a CSV template file with headers which should be
                     exported.
      --logLevel     Logging level: error, warn, info or verbose.
                                                               [default: \\"info\\"]
      --logFile      Path to file where to save logs.
                                           [default: \\"inventories-exporter.log\\"]
"
`;
